opm/grid/cpgrid/CpGridData.hpp

```diff
@@ -1123,22 +618,29 @@ class CpGridData
     /** @brief The boundary ids. */
     cpgrid::EntityVariable<int, 1> unique_boundary_ids_;
     /** @brief The index set of the grid (level). */
-    cpgrid::IndexSet* index_set_;
+    std::shared_ptr<cpgrid::IndexSet> index_set_;
     /** @brief The internal local id set (not exported). */
-    const cpgrid::IdSet* local_id_set_;
+    std::shared_ptr<const cpgrid::IdSet> local_id_set_;
     /** @brief The global id set (used also as local id set). */
-    LevelGlobalIdSet* global_id_set_;
+    std::shared_ptr<LevelGlobalIdSet> global_id_set_;
```

Comment:
Turning index_set_ into `unquire_ptr<>` instead of `shared_ptr<>` was quite straightforward. For `local_id_set_` and `global_id_set_`,  more work is required. Maybe we can postpone this. 

About unique pointers, we might need to use std::move() instead of copying, and then the object we moved from stays in undefined behavior, unless we always remember to initialize it again.

Reply:
> About unique pointers, we might need to use `std::move()` instead of copying

True.  Unique pointers have "unique" ownership of some resource (typically a dynamically allocated object).  When you wish to transfer that ownership to something else you have to `move()` it.  If you just want to look at the underlying object managed by a unique pointer `up`  you will typically write `*up` or `up->` or even `up.get()` instead.  The latter accesses the underlying pointer directly.

> and then the object we moved from stays in undefined behavior, unless we always remember to initialize it again.

That's not entirely accurate.  `std::unique_ptr<>` guarantees that the pointer is "nulled out" when moved from so if we have
```C++
std::unique_ptr<T> up1 = std::make_unique<T>(...);
// ...
std::unique_ptr<T> up2 = std::move(up1);
```
then `up1` is null/empty after the `std::move()` on line 2.  You can check for that condition using logic of the form
```C++
if (!up1) {
    // up1 was moved from
}
```
or maybe
```C++
if (up1 == nullptr) {
    // up1 was moved from
}
```
I usually prefer that second option, personally, because I think it's more explicit, but this is in some sense a question of "personal taste"-

Reply:
Many thanks for the detail explanation!